### PR TITLE
JAMES-3359 Allow mailbox parentId updates when have child

### PR DESCRIPTION
**Pull request by chibenwa**
Looping prevention is enforced by ensuring the new parentId is not one of the childs

### CODE REVIEW

**Review by rouazana — COMMENTED** [review]
the main remaining issue is concurrency. What happens if the user concurrently modifies mailboxes in such a way that between the moment you read it and the moment you write the change a loop is created?

**Comment by chibenwa** [comment]
> the main remaining issue is concurrency. What happens if the user concurrently modifies mailboxes in such a way that between the moment you read it and the moment you write the change a loop is created?

Obviously there's no way to defend against that, especially on top of Cassandra where we cannot rely on transaction.

**Comment by rouazana** [comment]
> Obviously there's no way to defend against that, especially on top of Cassandra where we cannot rely on transaction.

For me that's the reason we did not implement mailbox moving before.

If we do like this, we break the contract that there cannot be any loop and we have to add safe-guards when this happens.

**Comment by chibenwa** [comment]
Well technically, `mailboxManager.renameMailbox` blindly follows the decision taken before, and only accept a mailboxPath.

A mailbox path is by definition a "flat" representation of the path to a given mailbox in the mailbox tree.

Our stored data structure is loop proofed.

We could argue that a concurrency scenario can make the mailboxpath generation flacky though...

**Comment by chibenwa** [comment]
Concurrency leads to the following mailboxes to be created:

```
mailboxId2 -  New path: #private:bob@domain.tld:mailbox1.mailbox2
mailboxId1 -  New path: #private:bob@domain.tld:mailbox2.mailbox1
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 1]
```diff
@@ -5941,6 +5943,159 @@ trait MailboxSetMethodContract {
          |}""".stripMargin)
   }
 
+  @RepeatedTest(100)
+  def concurrencyChecksUponParentIdUpdate(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5947]: Not sure to get this test... You are trying to assign mailbox1 as parent of mailbox2 and mailbox2 as parent of mailbox1 concurrently right? Even with your loop detection logic I would not expect there is no update at all in both mailboxes (that's what you assert right?). Even with the comment at the end I don't understand well what is going on

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5947]: > You are trying to assign mailbox1 as parent of mailbox2 and mailbox2 as parent of mailbox1 concurrently right?

Yes

> Even with your loop detection logic I would not expect there is no update at all in both mailboxes (that's what you assert right?).

Both mailbox get's updated, data race happens ~50% of the time.

> Even with the comment at the end I don't understand well what is going on

What's going on is that the data is still accessible, nothing "blew up".

**Review by Arsnael — APPROVED** [review]

**Comment by rouazana** [comment]
I did not read the implementation, but I'm not sure I should have to understand your change and your test.

What happens for example in the case of an A/B/C loop? Is it also handled?

**Comment by chibenwa** [comment]
Our code only store **MailboxPath** that are flat, represents the access to a leaf of the mailbox branch.

We can thus **be sure than no loops are ever stored**.

So even in the face of a loop, the impact is very low - as data can't be screwed up severely on the server side.

As explained above the consequences of data races are:
 - An unexpected parent mailbox (different than the one specified by the user)
 - Or a missing parent mailbox - jmap will render it as a top level mailbox.

The parentId can be updated BTW hence the user can solve unexpected (unlikely!) data race issues by himself.

Hence thats why I concider acceptable.

**Comment by rouazana** [comment]
> Hence thats why I concider acceptable.

Seems fair, thank you for the nice explanation.

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**